### PR TITLE
Add tests for /bech32, 100% test coverage

### PR DESCRIPTION
**Pull request by dcousens**
~~Pre-cursor to https://github.com/bitcoinjs/bech32/pull/8/files?~~
~~Just getting the test fixtures in so I can understand why https://github.com/bitcoinjs/bech32/issues/7 was an issue.~~

Alternative to #8 (albeit missing the removal of the SegWit address exports)

### CODE REVIEW

File: bech32.js [hunk 1]
```diff
@@ -110,8 +110,7 @@ function decode (str) {
     }
   }
 
-  assert.equal(chk & 0x1, 1)
-
+  if (chk !== 1) throw new Error('Invalid checksum for ' + str)
```
Comment (dcousens) [inline bech32.js]: @afk11 truly, I have no idea why it was `& 1`

**Comment by dcousens** [comment]
Everything looks good except the `11qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqc8247j` fixture has a weird edge case where its data doesn't convert back and forth... at all.

See the reference implementation output:
```
var bech32 = require('./bech32')
var sw = require('./segwit_addr')
var f = '11qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqc8247j'

var r = bech32.decode(f)
var z = bech32.encode(r.hrp, r.data)
var bits = r.data

console.log(z, z === f)

console.log('padded ', JSON.stringify(sw.convertbits(bits, 5, 8, true)))
console.log('nopadd ', JSON.stringify(sw.convertbits(bits, 5, 8, false)))

var rebits2 = sw.convertbits(sw.convertbits(bits, 5, 8, false), 8, 5, false)
var rebits3 = sw.convertbits(sw.convertbits(bits, 5, 8, true), 8, 5, true)
console.log(bech32.encode(r.hrp, rebits2), z === f)
console.log(bech32.encode(r.hrp, rebits3), z === f)
```

**Comment by dcousens** [comment]
@sipa I suppose the 5-bit words returned from `11qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqc8247j` would always have no meaning post `convertbits`?

**Comment by dcousens** [comment]
Disregard,  it appears the `11qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqc8247j` works if you decode it with no padding, then re-encode it **with** padding.

Tricky @sipa keeping us on our toes.

**Comment by dcousens** [comment]
@afk11 did you have any other invalid checksum tests for #7 ?
Otherwise I'm going to close it in favour of this.

File: bech32.js [hunk 2]
```diff
@@ -94,28 +93,49 @@ function decode (str) {
     chk = polymodStep(chk) ^ (c & 0x1f)
   }
 
-  // NOTE: zero-fill required
-  let result = Buffer.alloc(bitData.length - 6)
+  let result = []
 
   for (let i = 0; i < bitData.length; ++i) {
-    let cv = bitData.charCodeAt(i)
-    assert.equal(cv & 0x80, 0)
```
Comment (dcousens) [inline bech32.js]: This is unnecessary,  either it is in `ALPHABET_MAP`, or it isn't

File: bech32.js [hunk 3]
```diff
@@ -94,28 +93,49 @@ function decode (str) {
...
   }
 
-  assert.equal(chk & 0x1, 1)
+  if (pad) {
+    if (bits > 0) {
+      result.push((value << (outBits - bits)) & maxV)
+    }
+  } else {
+    if (bits >= inBits) throw new Error('Bad bits 1')
+    if ((value << (outBits - bits)) & maxV) throw new Error('Bad bits 2')
```
Comment (dcousens) [inline bech32.js]: Haven't looked into the logic of this function yet to know the real meaning of these errors

**Comment by dcousens** [comment]
My bet is `11qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqc8247j` is encoded in the same a SEGWIT address is,  and that the `0` on the front is a `version: 0` byte... 100%.

**Comment by afk11** [comment]
I'll check the fixture later, but I really would rather we split up encode/decode more. 

keeping them in one function makes it harder to see what is happening, plus duplicating the same piece of code (as we have right now for hrpExpand in encode and decode) means we're highly likely to introduce incompatibilities bugs as we continue.

**Comment by dcousens** [comment]
@afk11 I'm not against that at all, for the record

File: bech32.js (hunk 3 above)
Comment (afk11) [inline bech32.js]: I think it's checking if bits weren't completely consumed, or if there was excessive padding.. sounds like we want a test case :)

**Comment by afk11** [comment]
Makes sense, sorry!

**Comment by dcousens** [comment]
@afk11 for reference,  I used the C code as reference: https://github.com/sipa/bech32/blob/master/ref/c/segwit_addr.c

File: test/fixtures.json [hunk 4]
```diff
@@ -53,32 +53,61 @@
...
         "string": "split1checkupstagehandshakeupstreamerranterredcaperred2y9e3w",
         "prefix": "split",
-        "hex": "c5f38b70305f519bf66d85fb6cf03058f3dde463ecd7918f2dc743918f2d"
+        "hex": "c5f38b70305f519bf66d85fb6cf03058f3dde463ecd7918f2dc743918f2d",
+        "words": [24,23,25,24,22,28,1,16,11,29,8,25,23,29,19,13,16,23,29,22,25,28,1,16,11,3,25,29,27,25,3,3,29,19,11,25,3,3,25,13,24,29,1,25,3,3,25,13]
       }
     ],
     "invalid": [
+      {
+        "string": "A12UEL5L",
```
Comment (sipa) [inline test/fixtures.json]: Uppercase Bech32 is valid.

File: bech32.js [hunk 5]
```diff
@@ -94,28 +94,56 @@ function decode (str) {
...
+}
+
+function toWords (bytes) {
+  return convert(bytes, 8, 5, true)
+}
 
-  return { prefix, bitData: result }
+function fromWords (words) {
+  return convert(words, 5, 8, false)
 }
```
Comment (dcousens) [inline bech32.js]: @afk11 `fromWords` does not pad

File: bech32.js [hunk 6]
```diff
@@ -94,28 +94,56 @@ function decode (str) {
...
+    if (bits >= inBits) throw new Error('Bad bits 1')
+    if ((value << (outBits - bits)) & maxV) throw new Error('Bad bits 2')
   }
 
-  assert.equal(chk & 0x1, 1)
+  return result
+}
+
+function toWords (bytes) {
+  return convert(bytes, 8, 5, true)
```
Comment (dcousens) [inline bech32.js]: `toWords` does

**Comment by dcousens** [comment]
@afk11 should be good to merge,  and then we can remove the address logic to `bitcoinjs` and have this purely be encoding

File: test/bech32.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  tape(`decode ${f.string}`, (t) => {
+    t.plan(1)
+
+    t.same(bech32.decode(f.string), {
+      prefix: f.prefix,
+      words: f.words
+    })
+  })
+
+  tape(`fails for ${f.string} with 1 bit flipped`, (t) => {
```
Comment (dcousens) [inline test/bech32.js]: Added these from #8 ,  closing #8 now

**Comment by dcousens** [comment]
[BIP173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki) is liberal in that it accepts uppercase XOR lowercase:
> Decoders MUST accept both uppercase and lowercase strings, but not mixed case. The lowercase form is used when determining a character's value for checksum purposes. For presentation, lowercase is usually preferable, but inside QR codes uppercase SHOULD be used, as those permit the use of alphanumeric mode, which is 45% more compact than the normal byte mode.

But as seen in 4eff274,  this will mean the `encode(decode(x...)) === x...` assertion won't hold.
I'd be tempted to allow that mixed case at the `bitcoinjs-lib` level,  but the BIP specifies this directly for the Bech32 encoding format.

We only ever output lowercase,  we could add a NOTE to the README about users enforcing `.toLowerCase` on their inputs to be BIP compliant?
Otherwise, 4eff274 is the [lame] reality.

**Comment by dcousens** [comment]
@afk11 I added exceptions and got the coverage over the threshold... now it should be good to go.

File: test/fixtures.json [hunk 8]
```diff
@@ -39,46 +39,131 @@
...
+        "string": "tb1qrp33g0q5c5txsp9arysrx4k6zdkfs4nce4xj0gdcccefvpysxf3q0sL5k7",
+        "exception": "Mixed-case string"
+      },
+      {
+        "string": "tb1pw508d6qejxtdg4y5r3zarqfsj6c3",
+        "exception": "Bad bits 1"
+      },
+      {
+        "string": "tb1qrp33g0q5c5txsp9arysrx4k6zdkfs4nce4xj0gdcccefvpysxf3pjxtptv",
+        "exception": "Bad bits 2"
```
Comment (dcousens) [inline test/fixtures.json]: We probably want insightful exception messages here...

File: index.js [hunk 9]
```diff
@@ -1,59 +1,36 @@
...
-  bitData.unshift(version)
+  let words = bech32.toWords(program)
+  words.unshift(version)
 
-  return bech32.encode(prefix, bitData)
+  return bech32.encode(prefix, words)
 }
 
-function decode (expectedPrefix, string) {
+function decode (string, expectedPrefix) {
```
Comment (dcousens) [inline index.js]: reversed this, it now matches what I we use elsewhere

File: test/fixtures.json [hunk 10]
```diff
@@ -39,46 +39,131 @@
...
-      "bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t5",
-      "BC13W508D6QEJXTDG4Y5R3ZARVARY0C5XW7KN40WF2",
-      "bc1rw5uspcuh",
-      "bc10w508d6qejxtdg4y5r3zarvary0c5xw7kw508d6qejxtdg4y5r3zarvary0c5xw7kw5rljs90",
-      "BC1QR508D6QEJXTDG4Y5R3ZARVARYV98GJ9P",
-      "tb1qrp33g0q5c5txsp9arysrx4k6zdkfs4nce4xj0gdcccefvpysxf3q0sL5k7",
-      "tb1pw508d6qejxtdg4y5r3zarqfsj6c3",
-      "tb1qrp33g0q5c5txsp9arysrx4k6zdkfs4nce4xj0gdcccefvpysxf3pjxtptv"
+      {
+        "ignore": true,
```
Comment (dcousens) [inline test/fixtures.json]: ignored, as this library doesn't care about network

**Comment by dcousens** [comment]
Added PR to fix reference implementation for ASCII characters that aren't in the range [33,126] https://github.com/sipa/bech32/pull/24

File: index.js [hunk 11]
```diff
@@ -1,59 +1,36 @@
...
   let result = bech32.decode(string)
-  assert.equal(result.prefix, expectedPrefix)
+  let prefix = result.prefix
+  let words = result.words
+  if ((expectedPrefix !== undefined) && (prefix !== expectedPrefix)) throw new Error('Unexpected prefix')
 
-  assert((result.bitData.length > 0) && (result.bitData.length < 66))
```
Comment (dcousens) [inline index.js]: Need to re-add these,  but,  test cases are required

File: test/fixtures.json [hunk 12]
```diff
@@ -50,18 +50,30 @@
       },
       {
         "string": "BC13W508D6QEJXTDG4Y5R3ZARVARY0C5XW7KN40WF2",
+        "prefix": "bc",
```
Comment (dcousens) [inline test/fixtures.json]: Note, `prefix` provided **must** be lowercase... as per the reference implementations too

**Comment by dcousens** [comment]
Mega-squashed. 
Tests added in https://github.com/bitcoinjs/bech32/pull/9/commits/b777532ce4e7416e7d2499c49c468dbb7afcd870

Implementation fixed in https://github.com/bitcoinjs/bech32/pull/9/commits/32223ab8bde1537de5ea31df9f929074652ed640

Optimizations there-after except for a1f39c3, which requires independent discussion.

**Comment by afk11** [comment]
@dcousens I just .toLowerCase in the assertion..

sweet, will review soon

**Comment by dcousens** [comment]
@afk11 the BIP is quite specific about supporting upper case for QR codes (compact),  but I'd still personally enforce `encode(decode(x)) == x`... what are your thoughts @afk11?

**Comment by afk11** [comment]
Probably better to just allow it

**Comment by dcousens** [comment]
Can we merge within 24 hrs? @afk11

File: bech32.js [hunk 13]
```diff
@@ -21,33 +20,34 @@ function polymodStep (pre) {
...
+  return chk
+}
 
-  for (let i = 0; i < bitData.length; ++i) {
-    let x = bitData[i]
-    assert.equal(x >> 5, 0)
+function encode (prefix, words) {
+  // too long?
+  if ((prefix.length + 7 + words.length) > 90) throw new TypeError('Exceeds Bech32 maximum length')
+  prefix = prefix.toLowerCase()
```
Comment (afk11) [inline bech32.js]: Reviewed for ways to remove this, unfortunately it's not straightforward without a for loop..

Comment (dcousens) [inline bech32.js]: I know,  we'll optimize it over time

**Pull request merged**